### ISSUE TITLE
gRPC status code header destroyed on error

### ISSUE THREAD

**Issue by hectim** [improbable-eng/grpc-web#225]
Been using this grpc-web client on a project for a while, but recently started handling specific grpc errors on the front-end. Our requests hit an Envoy proxy, which does its own httpToGrpc status conversion, but in the grpc-web-client code specifically we are also losing the grpc-status header. 

Through some local testing, it looks like the problem comes from https://github.com/improbable-eng/grpc-web/blob/master/ts/src/client.ts#L99
```
const code = httpStatusToCode(status);
this.props.debug && debug("onHeaders.code", code);
const gRPCMessage = headers.get("grpc-message") || [];
this.props.debug && debug("onHeaders.gRPCMessage", gRPCMessage);
if (code !== Code.OK) {
  const statusMessage = this.decodeGRPCStatus(gRPCMessage[0]);
  this.rawOnError(code, statusMessage);
  return;
}
```

My problem could be solved if instead of passing the newly created `code` to rawOnError, it checked for a grpc-status header first and used that value before defaulting to the new `code`.  Something like: 
```
const code = parseInt(headers.get("grpc-status"), 10) || httpStatusToCode(status);
```
This seems to be the simplest way to fix it, although there may be reasons why this is unwanted. 

Option 2: If we made sure the headers were being passed correctly, we could just pull them off the response. Right now when it throws an error it makes a `new Metadata()` and hands that off, but it that isn't very helpful. 
```
rawOnError(code: Code, msg: string) {
    this.props.debug && debug("rawOnError", code, msg);
    if (this.completed) return;
    this.completed = true;
    this.onEndCallbacks.forEach(callback => {
      detach(() => {
        callback(code, msg, new Metadata());
      });
    });
  }
```

If instead, rawOnError took an optional third parameter of the headers 
```
rawOnError(code: Code, msg: string, headers?: Metadata) {
```
and passed those (if they exist) to the callback, we could accomplish essentially the same thing. 
```
callback(code, msg, headers || new Metadata());
```

I could submit a PR for this anytime, just wanted to know if there were any extra thoughts about potential hiccups or problems with either of these. Personally, I would opt for the first option for simplicity.

**Comment by gunn4r** [comment]
@easyCZ I'm working with @hectim on this and will be putting together the actual PR. Thanks a bunch!

**Comment by gunn4r** [comment]
@easyCZ After digging through this more, it appears the tests that are testing "error" cases are not actually ever hitting the error case here:

```if (code !== Code.OK) {
        const statusMessage = this.decodeGRPCStatus(gRPCMessage[0]);
        this.rawOnError(code, statusMessage);
        return;
      }```

Because the test server returns 200 responses, and with the current way the responses are handled that gets converted into a grpc status of 0 just above the if statement:

`const code = httpStatusToCode(status);`

So in the tests, while the test server returns a grpc error status, the http status is 200, thus bypassing the if statement. Many tests test for "headers / trailers" which are never available inside the error case.  So it looks like to me that the tests are giving false positives.

Are you guys willing to address this issue?  Thanks :) 👍

**Comment by gunn4r** [comment]
If I make the original proposed changes, and also call `rawOnHeaders` when there is a grpc error code then every thing works peachy and all the tests pass. Basically move the `this.rawOnHeaders(headers);` from below the if statement checking the grpc code to above it. I'm not seeing any complications from that but I could be wrong.  It WOULD mean that onHeaders callbacks would get called on errors as well as non errors but with how you guys are currently using it that seems to be the case anyway (since you guys are returning http 200s all the time regardless of the grpc status).  Let me know what you think @easyCZ.

This is what that change looks like:

```
onTransportHeaders(headers: Metadata, status: number) {
    this.props.debug && debug("onHeaders", headers, status);

    if (this.closed) {
      this.props.debug && debug("grpc.onHeaders received after request was closed - ignoring");
      return;
    }

    if (status === 0) {
      // The request has failed due to connectivity issues. Do not capture the headers
    } else {
      this.responseHeaders = headers;
      this.props.debug && debug("onHeaders.responseHeaders", JSON.stringify(this.responseHeaders, null, 2));

      const gRPCStatus: Code = parseInt(headers.get("grpc-status")[0], 10);

      const code: Code = gRPCStatus >= 0 ? gRPCStatus : httpStatusToCode(status);

      this.props.debug && debug("onHeaders.code", code);

      const gRPCMessage = headers.get("grpc-message") || [];

      this.props.debug && debug("onHeaders.gRPCMessage", gRPCMessage);

      this.rawOnHeaders(headers);

      if (code !== Code.OK) {
        const statusMessage = this.decodeGRPCStatus(gRPCMessage[0]);
        this.rawOnError(code, statusMessage, headers);
      }
    }
  }
```

```
  rawOnError(code: Code, msg: string, headers?: Metadata) {
    this.props.debug && debug("rawOnError", code, msg);
    if (this.completed) return;
    this.completed = true;
    this.onEndCallbacks.forEach(callback => {
      detach(() => {
        callback(code, msg, headers || new Metadata());
      });
    });
  }
```

**Comment by jonny-improbable** [comment]
I believe this was fixed by #226.
